### ISSUE TITLE
Problems with DNS resolution of newly created records

### ISSUE THREAD

**Issue by adamwalach** [kyma-project/test-infra#2940]
**Description**

On multiple prow jobs we observe problems with DNS resolution, for example: https://storage.googleapis.com/kyma-prow-logs/pr-logs/pull/kyma-project_kyma/9850/pre-master-kyma-gke-integration/1324283142663573504/build-log.txt
```
Trying to resolve *.gkeint-pr-9850-ecu27oxj54.a.build.kyma-project.io.
2020/11/05 09:53:19 UTC [INFO] *************************************************************************************
2020/11/05 09:53:19 UTC [INFO] * Debugging DNS issues
2020/11/05 09:53:19 UTC [INFO] *************************************************************************************
2020/11/05 09:53:19 UTC
Cannot resolve *.gkeint-pr-9850-ecu27oxj54.a.build.kyma-project.io. to expected IP_ADDRESS: 34.91.211.197.
2020/11/05 09:53:20 UTC [INFO] *************************************************************************************
2020/11/05 09:53:20 UTC [INFO] * CLEANUP
2020/11/05 09:53:20 UTC [INFO] *************************************************************************************
2020/11/05 09:53:20 UTC [ERROR] AN ERROR OCCURED! Take a look at preceding log entries.
```

This problem happens every few weeks, sometimes days and lasts for a few hours. We could not find any pattern, any correlation to any action that we did on our infrastructure. 
Outcome is always the same:
- inability to resolve dns entries that we created few minutes before
- querying directly 4 dns servers responsible for our subdomains gives mixed results, sometimes even after 30mins only one of the servers returns correct answer
- all jobs that need external domain are failing (mainly gke*)

What we already tried:
- kubedns restart 
`kubectl -n kube-system scale deployment kube-dns --replicas=0`
- adding [stubdomains](https://github.com/kyma-project/test-infra/blob/b0cbfeed74b1ddb28572390b42b04ec9f9f17a0c/prow/scripts/resources/kube-dns-stub-domains-patch.yaml#L11) section to our workload and test clusters configuration
- asking gcp support
When it happens it basically blocks our development process - jobs are failing, PRs are blocked.

**Expected result**

Updates or new DNS entries should be visible inside of a pod in few seconds/minutes

**Actual result**

Sometimes records are stale/not present at all

**Steps to reproduce**

Looks like a random behaviour

**Comment by adamwalach** [comment]
The only solution that we have for now is to replace CloudDNS with Route53. For no apparent reason it works much faster (few seconds).

**Comment by dekiel** [comment]
To troubleshoot and get more data we replaced default gke kube-dns deployment with same setup but with enabled logging of queries. Additionally we added dns debugging commands to script responsbile for creating and validating dns records. In debugging commands we query directly all dns authoritative servers for our domain. We do print /etc/resolv/conf file and check what is kube-dns cluster IP. We do query kube-dns service endpoints directly. From collected data we learned follwoing.
1. Valid A record appear first one of google dns servers.
2. Even when this record appear on one server. Resolution through kube-dns still failing.
3. Negative caching on dnsmasq is disabled.
4. We see queries reach kube-dns dnsmasq and are forwarded to all upstream dns servers provided within stubdomain. There is no mention in logs about replies send back to pod.

We have raised gcp support ticket for this issue.

**Comment by dekiel** [comment]
When using kube-dns deployment as a k8s cluster DNS all requests for non cluster domains are routed to upstream servers by dnsmasq which is a frontend cache and forwarder for real kube-dns container. According to the dnsmasq documentation and [this post](http://lists.thekelleys.org.uk/pipermail/dnsmasq-discuss/2009q3/003295.html), dnsmaq will query all upstream servers defined under stub domain entry in parallel and send first received reply to the client. Because of that, dnsmasq will reply with NXDOMAIN which is a valid DNS response even when one of upstream servers hold IP for resolved record, but response with that IP, dnsmasq will get after the response with NXDOMAIN.

However during our tests, we did notice that resolving IPs from our workstations works each time, even when kube-dns doesn't. I think this is caused by different way of getting records from DNS than send DNS query to NS servers for queried domain. Most probably system behind public DNS servers use some kind API to get zone records which will overcome NS servers replication delay. To test this assumption we set upstream servers for our domains to 8.8.8.8 and 8.8.4.4. If we will not get any DNS resolution failures for two weeks, I will consider problem solved.

If this will not solve the problem, we might consider replacing kube-dns with core-dns and using clouddns plugin which will use cloudDNS API to get zone records.
